### PR TITLE
Dictionary methods

### PR DESCRIPTION
**Pull request by lummax**

### CODE REVIEW

**Comment by meh** [comment]
I'm not that sure about the `From<HashMap<&str, &str>>` implementation, especially since in general it would be an `HashMap<String, String>` instead.

**Comment by lummax** [comment]
I actually tried som `Into` magic but failed (or did not try hard enough).

I'm not sure about the general use case. I think it's mostly stuff where you build a `HashMap<&'static str, String>` with string literals as keys and values build with `format!()`.

But either implementation is fine (`&str` or `String`) or both. Should I change it?

**Comment by lummax** [comment]
Rebased and changed the implementation to `From<Hashmap<String, String>>`.

**Comment by meh** [comment]
This pull request made me rethink how dictionaries are handled, but I need to know how they're actually used by ffmpeg.

For instance, the metadata `Dictionary` found in various places, is it possible to modify it in-place as far as the usage of ffmpeg is concerned?

We could split it in `DictionaryRef`, `DictionaryMut` and `Dictionary`.

- `DictionaryRef` is an immutable reference (non-owned) to an `AVDictionary`.
- `DictionaryMut` is a mutable reference (non-owned) to an `AVDictionary`.
- `Dictionary` is an owned `AVDictionary`.

Basically every method taking a `Dictionary` now would take a `DictionaryRef`.

Then we could have `metadata_mut` where you want to actually modify the contents of some metadata, which would return a `DictionaryMut`.

Then both `DictionaryRef` and `DictionaryMut` could implement `ToOwned`, which would create a `Dictionary` out of them.

I think I'll go this way.

**Comment by lummax** [comment]
As I understand it there are two use cases for `AVDictionary` in ffmpeg:

1) Getting medata. Here a `DictionaryRef` would suffice.
2) Setting metadata. I think an owned `Dictionary` where you transfer ownership on setting the metada could be a way. But that would restrict setting the medata to one place instead of setting various metadata keys all over the place.
3) Passing additional arguments. In this case the ffmpeg API expects an `AVDictionary` where some options are set. And it will (as I understand it) clear the dictionary and set some options itself (as an output parameter).

For both 1 and 2 an owned `Dictionary` is needed, as the user constructs it.

(I had some weird bugs with this. For example `outstream.metadata().set("key", "value")` would not work as some internal pointer got screwed up. That's why I would refrein from the `DictionaryMut` variant.)

**Comment by meh** [comment]
Ok, I merged all your commits except the `From` one, I implemented `FromIterator` instead, which allows us to create a `Dictionary` from a `collect()` call.
